File: csv_to_sql.php
<?php

// COMO USAR: 
// source ~/.bashrc; php csv_to_sql.php (de forma infetariva) ou
// source ~/.bashrc; php csv_to_sql.php "seu_arquivo.csv" (conversão direta)


// CONFIGURAÇÕES:
$SGBD=         "postgres";          // "postgres" | "mysql"
$rView=         5;                  // Linhas em preview
$batchSize=     500;                // Blocos de insert
$outpuPath=     "/app/backups/";    // pasta de saída



function normalize_name($name) {
    $name = strtolower(trim($name));
    $name = preg_replace('/[^\w\s]/', '', $name); // Trata acentos e símbolos
    $name = preg_replace('/[\s\-\.]+/', '_', $name); // Trata espaços e pontuações por underline
    return $name;
}

function detect_delimiter($filePath) {
    $delimiters = [",", ";", "\t", "|"];
    $line = fgets(fopen($filePath, 'r'));
    $results = [];
    foreach ($delimiters as $delimiter) {
        $fields = str_getcsv($line, $delimiter);
        $results[$delimiter] = count($fields);
    }
    arsort($results);
    return array_key_first($results);
}

function detect_column_type($values, $SGBD) {
    $isNumeric = true;
    $isText = false;
    foreach ($values as $value){
        if (is_numeric($value)) continue;
        $isNumeric = false;
        if (strlen($value) > 0) $isText = true;
    }
    if ($SGBD == "mysql"){ return $isNumeric ? 'INT' : ($isText ? 'VARCHAR(255)' : 'TEXT'); // MySQL
    }elseif($SGBD == "postgres"){ return $isNumeric ? 'INTEGER' : ($isText ? 'VARCHAR(255)' : 'TEXT'); // Postgres
    }else{ return 'TEXT'; // fallback seguro (em ultimo caso, como segurança adicional)
    }
}

function prompt($message){
    echo $message;
    return trim(fgets(STDIN));
}

if ($argc === 1) {
    $csvFile = prompt("Digite o caminho do arquivo CSV: ");
    if (!file_exists($csvFile) || !is_readable($csvFile)) {
        echo "Erro: Arquivo não encontrado ou não legível.\n";
        exit(1);
    }
    $delimiter = detect_delimiter($csvFile);
    $handlePreview = fopen($csvFile, 'r');
    echo "\n\n\033[37;44m PRÉ-VISUALIZAÇÃO DOS DADOS \033[0m (máximo $rView linhas):\n";
    $count = 0;
    while (($line = fgetcsv($handlePreview, 0, $delimiter)) !== false && $count < $rView) {
        echo implode(" | ", $line) . "\n";
        $count++;
    }
    fclose($handlePreview);
    $tableNameInput = prompt("\n\033[37;44m DIGITE O NOME DA TABELA SQL \033[0m \033[37m\n(ou pressione Enter para usar o nome do arquivo):\033[0m ");
    $tableName = $tableNameInput ? normalize_name($tableNameInput) : normalize_name(pathinfo($csvFile, PATHINFO_FILENAME));
} elseif ($argc === 2) {
    $csvFile = $argv[1];
    $tableName = normalize_name(pathinfo($csvFile, PATHINFO_FILENAME));
    $delimiter = detect_delimiter($csvFile);
} elseif ($argc === 3) {
    $csvFile = $argv[1];
    $tableName = normalize_name($argv[2]);
    $delimiter = detect_delimiter($csvFile);
} else {
    echo "USO: php csv_to_sql.php \"caminho/para/arquivo.csv\" \"nome da tabela\"\n";
    echo "OBS: O uso de aspas é obrigatório.\n";
    exit(1);
}
if (!file_exists($csvFile) || !is_readable($csvFile)) {
    echo "Erro: Arquivo não encontrado ou não legível: $csvFile\n";
    exit(1);
}
$start = microtime(true);
$SGBD = mb_strtolower($SGBD, 'UTF-8'); // garante padrão minuscula
$datePrefix = date("d.m.Y");
$outputFile = "{$outpuPath}{$datePrefix}_{$tableName}.sql";
$handle = fopen($csvFile, 'r');
$columns = fgetcsv($handle, 0, $delimiter);
if (!$columns) {
    echo "Erro: Cabeçalho vazio ou inválido.\n";
    exit(1);
}
$columns = array_map('normalize_name', $columns);

// Seleciona tipos de dados/colunas
$previewRows = [];
$count = 0;
while (($row = fgetcsv($handle, 0, $delimiter)) !== false && $count < 10) {
    $previewRows[] = $row;
    $count++;
}
fclose($handle);
$columnTypes = [];
foreach ($columns as $index => $col) {
    $columnValues = array_column($previewRows, $index);
    $columnTypes[$col] = detect_column_type($columnValues, $SGBD);
}

$out = fopen($outputFile, 'w');
fwrite($out, "-- Gerado a partir de $csvFile \n-- Padrão: {$SGBD}\n\n");

// CREATE TABLE (se não existir)
switch($SGBD){
    case "mysql":
        $tableSQL = "CREATE TABLE IF NOT EXISTS `$tableName` (\n";
        $tableSQL .= "  `id` INT AUTO_INCREMENT PRIMARY KEY,\n";
        foreach ($columns as $col){
            if ($col !== 'id'){
                $type = $columnTypes[$col] ?? 'VARCHAR(255)';
                $tableSQL .= "  `$col` $type,\n";
            }
        }
    break;
    case "postgres":
        $tableSQL = "CREATE TABLE IF NOT EXISTS \"$tableName\" (\n";
        $tableSQL .= "  \"id\" SERIAL PRIMARY KEY,\n";
        foreach ($columns as $col){
            if ($col !== 'id'){
                $type = $columnTypes[$col] ?? 'VARCHAR(255)';
                $tableSQL .= "  \"$col\" $type,\n";
            }
        }
    break;
}
$tableSQL = rtrim($tableSQL, ",\n") . "\n);\n\n";
fwrite($out, $tableSQL);

// INSERT INTO
$handle = fopen($csvFile, 'r');
$columns = array_map('normalize_name', fgetcsv($handle, 0, $delimiter));
$recordCount = 0;
$insertBatch = [];
switch($SGBD){
    case "mysql":
        while (($row = fgetcsv($handle, 0, $delimiter)) !== false){
            if (count($row) !== count($columns)) continue;
            $escaped = array_map(function ($v) {
                return is_numeric($v) ? $v : "'" . str_replace("'", "\'", trim($v)) . "'";
            }, $row);
            $insertBatch[] = "(" . implode(", ", $escaped) . ")";
            $recordCount++;
            if (count($insertBatch) >= $batchSize) {
                fwrite($out, "BEGIN; \nINSERT INTO `$tableName` (`" . implode("`, `", $columns) . "`) VALUES\n");
                fwrite($out, implode(",\n", $insertBatch) . ";\nCOMMIT; \n\n");
                $insertBatch = [];
            }
        }
        if (!empty($insertBatch)) {
            fwrite($out, "BEGIN; \nINSERT INTO `$tableName` (`" . implode("`, `", $columns) . "`) VALUES\n");
            fwrite($out, implode(",\n", $insertBatch) . ";\nCOMMIT; \n\n");
        }
    break;
    case "postgres":
        while (($row = fgetcsv($handle, 0, $delimiter)) !== false){
            if (count($row) !== count($columns)) continue;
            $escaped = array_map(function ($v) {
                return is_numeric($v) ? $v : "'" . str_replace("'", "''", trim($v)) . "'";
            }, $row);
            $insertBatch[] = "(" . implode(", ", $escaped) . ")";
            $recordCount++;
            if (count($insertBatch) >= $batchSize) {
                fwrite($out, "BEGIN; \nINSERT INTO \"$tableName\" (\"" . implode("\", \"", $columns) . "\") VALUES\n");
                fwrite($out, implode(",\n", $insertBatch) . ";\nCOMMIT; \n\n");
                $insertBatch = [];
            }
        }
        if (!empty($insertBatch)) {
            fwrite($out, "BEGIN; \nINSERT INTO \"$tableName\" (\"" . implode("\", \"", $columns) . "\") VALUES\n");
            fwrite($out, implode(",\n", $insertBatch) . ";\nCOMMIT; \n\n");
        }
    break;
}

fclose($handle);
fclose($out);

$duration = microtime(true) - $start;
echo "\n\033[37;44m ✅ CONVERSÃO CONCLUÍDA \033[0m\n";
echo "Script {$SGBD} gerado em: \033[37m$outputFile\033[0m\n";
echo "Linhas processadas (dados): $recordCount\n";
printf("Todo o processo levou: %s%s\n\n\n", $duration > 59 ? ($f=gmdate("H:i:s",(int)$duration)) . " (" : '', number_format($duration, 2) . " segundos" . ($duration > 59 ? ')' : ''));

File: csv_to_sql.py
#!/usr/bin/env python3

# Biblioteca usadas
import os                       # os — Operações com o sistema de arquivos, como caminhos e verificação de existência
import sys                      # sys — Permite acesso aos argumentos da linha de comando e outras funções do sistema
import csv                      # csv — Leitura e escrita de arquivos CSV com suporte a diferentes delimitadores
import re                       # re — Lida com expressões regulares, como normalização de strings
from datetime import datetime   # datetime — Manipulação de datas e horas, gera prefixos e timestamps
from dateutil.parser import parse as parse_date  # dateutil.parser — Submódulo da dateutil (para manipulação de datas completas) que contém a função parse (identifica strings como datas)
import time                     # time — Mede duração de execução e fornece funções de tempo em segundos




# COMO USAR: 
# source ~/.bashrc; python csv_to_sql.py (de forma interativa) ou
# source ~/.bashrc; python csv_to_sql.py "seu_arquivo.csv" (conversão direta)
# source ~/.bashrc; python csv_to_sql.py "seu_arquivo.csv" "nome_da_tabela" (direta, com nome de tabela explícito)

# CONFIGURAÇÕES: 
SGBD=           "postgres"         # "postgres" | "mysql"
rView=          5                  # Linhas em preview
batchSize=      500                # Blocos de insert
outputPath=     "/app/backups/"    # pasta de saída





def normalize_name(name):
    name = name.strip().lower()
    name = re.sub(r'[^\w\s]', '', name)
    name = re.sub(r'[\s\-\.]+', '_', name)
    return name

def detect_delimiter(file_path):
    delimiters = [",", ";", "\t", "|"]
    with open(file_path, 'r', encoding='utf-8') as f:
        line = f.readline()
        results = {d: len(line.split(d)) for d in delimiters}
    return max(results, key=results.get)

def detect_column_type(values, sgbd):
    bool_set = {'true', 'false', 'yes', 'no', 'sim', 'não', '0', '1'}
    non_empty_values = [v.strip().lower() for v in values if v.strip()]
    type_balance = {'boolean': 0, 'date': 0, 'integer': 0, 'float': 0, 'text': 0}
    for v in non_empty_values:
        if v in bool_set:
            type_balance['boolean'] += 1
        elif v.lstrip('0') != v and len(v) > 1:  # Provável zeros à esquerda
            type_balance['text'] += 1
        elif v.replace(',', '.').replace('.', '', 1).isdigit():
            type_balance['float' if '.' in v or ',' in v else 'integer'] += 1
        elif re.match(r'^\d{4}-\d{2}-\d{2}$', v):  # Date (ISO)
            type_balance['date'] += 1
        else:
            try:
                parse_date(v, dayfirst=True)
                type_balance['date'] += 1
            except:
                type_balance['text'] += 1
    best_type = max(type_balance, key=type_balance.get) if len(non_empty_values) == 0 or type_balance[max(type_balance, key=type_balance.get)] / len(non_empty_values) >= 0.8 else 'text'
    return {
        'mysql': {'boolean': 'BOOLEAN', 'date': 'DATE', 'integer': 'INT', 'float': 'FLOAT', 'text': 'TEXT'},
        'postgres': {'boolean': 'BOOLEAN', 'date': 'DATE', 'integer': 'INTEGER', 'float': 'REAL', 'text': 'TEXT'}
    }.get(sgbd, {}).get(best_type, 'TEXT') # fallback adicional para segurança de tipos (previne perda de dados)

def prompt(message):
    return input(message).strip()

def preview_csv(file_path, delimiter, limit):
    with open(file_path, 'r', encoding='utf-8') as f:
        reader = csv.reader(f, delimiter=delimiter)
        print(f"\n\n\033[37;44m PRÉ-VISUALIZAÇÃO DOS DADOS \033[0m (máximo {limit} linhas):")
        for i, row in enumerate(reader):
            print(" | ".join(row))
            if i + 1 >= limit:
                break

def main():
    args = sys.argv
    if len(args) == 1:
        csv_file = prompt("Digite o caminho do arquivo CSV: ")
        if not os.path.exists(csv_file):
            print("Erro: Arquivo não encontrado.")
            return
        delimiter = detect_delimiter(csv_file)
        preview_csv(csv_file, delimiter, rView)
        table_name_input = prompt("\n\033[37;44m DIGITE O NOME DA TABELA SQL \033[0m \033[37m\n(ou pressione Enter para usar o nome do arquivo):\033[0m ")
        table_name = normalize_name(table_name_input) if table_name_input else normalize_name(os.path.splitext(os.path.basename(csv_file))[0])
    elif len(args) == 2:
        csv_file = args[1]
        table_name = normalize_name(os.path.splitext(os.path.basename(csv_file))[0])
        delimiter = detect_delimiter(csv_file)
    elif len(args) == 3:
        csv_file = args[1]
        table_name = normalize_name(args[2])
        delimiter = detect_delimiter(csv_file)
    else:
        print("USO: python csv_to_sql.py \"caminho/para/arquivo.csv\" \"nome da tabela\"")
        return

    if not os.path.isfile(csv_file):
        print(f"Erro: Arquivo não encontrado ou não legível: {csv_file}")
        return

    start_time = time.time()
    sgbd = SGBD.lower()
    date_prefix = datetime.now().strftime('%d.%m.%Y')
    output_file = os.path.join(outputPath, f"{date_prefix}_{table_name}.sql")
    with open(csv_file, 'r', encoding='utf-8') as f:
        reader = csv.reader(f, delimiter=delimiter)
        headers = next(reader)
        columns = [normalize_name(h) for h in headers]
        preview_rows = [row for _, row in zip(range(10), reader)]
    column_types = {}
    for idx, col in enumerate(columns):
        values = [row[idx] for row in preview_rows if idx < len(row)]
        column_types[col] = detect_column_type(values, sgbd)
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    with open(output_file, 'w', encoding='utf-8') as out:
        out.write(f"-- Gerado a partir de {csv_file} \n-- Padrão: {sgbd}\n\n")

        # CREATE TABLE
        if sgbd == "mysql":
            out.write(f"CREATE TABLE IF NOT EXISTS `{table_name}` (\n")
            out.write("  `id` INT AUTO_INCREMENT PRIMARY KEY,\n")
            for col in columns:
                if col != 'id':
                    out.write(f"  `{col}` {column_types[col]},\n")
        elif sgbd == "postgres":
            out.write(f"CREATE TABLE IF NOT EXISTS \"{table_name}\" (\n")
            out.write("  \"id\" SERIAL PRIMARY KEY,\n")
            for col in columns:
                if col != 'id':
                    out.write(f"  \"{col}\" {column_types[col]},\n")
        out.seek(out.tell() - 2)
        out.write("\n);\n\n")

        # INSERTs
        with open(csv_file, 'r', encoding='utf-8') as f:
            reader = csv.reader(f, delimiter=delimiter)
            next(reader)  # skip header
            insert_batch = []
            record_count = 0
            for row in reader:
                if len(row) != len(columns):
                    continue
                def escape(val):
                    val = val.strip()
                    if val.replace('.', '', 1).isdigit():
                        return val
                    if sgbd == 'mysql':
                        return "'" + val.replace("'", "\\'") + "'"
                    elif sgbd == "postgres":
                        return "'" + val.replace("'", "''") + "'"
                escaped = [escape(v) for v in row]
                insert_batch.append(f"({', '.join(escaped)})")
                record_count += 1
                if len(insert_batch) >= batchSize:
                    if sgbd == "mysql":
                        out.write("BEGIN;\nINSERT INTO `{}` (`{}`) VALUES\n".format(table_name, '`, `'.join(columns)))
                    elif sgbd == "postgres":
                        out.write("BEGIN;\nINSERT INTO \"{}\" (\"{}\") VALUES\n".format(table_name, "\", \"".join(columns)))
                    out.write(",\n".join(insert_batch) + ";\nCOMMIT;\n\n")
                    insert_batch = []

            if insert_batch:
                if sgbd == "mysql":
                    out.write("BEGIN;\nINSERT INTO `{}` (`{}`) VALUES\n".format(table_name, '`, `'.join(columns)))
                elif sgbd == "postgres":
                    out.write("BEGIN;\nINSERT INTO \"{}\" (\"{}\") VALUES\n".format(table_name, "\", \"".join(columns)))
                out.write(",\n".join(insert_batch) + ";\nCOMMIT;\n\n")

    duration = time.time() - start_time
    print("\n\033[37;44m ✅ CONVERSÃO CONCLUÍDA \033[0m")
    print(f"Script {sgbd} gerado em: \033[37m{output_file}\033[0m")
    print(f"Linhas processadas (dados): {record_count}")
    print(f"Todo o processo levou: {datetime.utcfromtimestamp(duration).strftime('%H:%M:%S')} ({duration:.2f} segundos)\n")






if __name__ == "__main__":
    main()
